mqtt/mqtt-policy/src/validator.rs

```diff
@@ -0,0 +1,322 @@
+        }
+    }
+}
+
+fn visit_definition(definition: &PolicyDefinition) -> Result<(), Error> {
+    let errors = definition
+        .statements()
+        .iter()
+        .flat_map(|statement| {
+            let statement_errors = visit_statement(statement);
+            let identity_errors = statement
+                .identities()
+                .iter()
+                .filter_map(|i| visit_identity(i).err());
+            let operation_errors = statement
+                .operations()
+                .iter()
+                .filter_map(|o| visit_operation(o).err());
+            let resource_errors = statement
+                .resources()
+                .iter()
+                .filter_map(|r| visit_resource(r).err());
+
+            statement_errors
+                .into_iter()
+                .chain(identity_errors)
+                .chain(operation_errors)
+                .chain(resource_errors)
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    if !errors.is_empty() {
+        return Err(Error::ValidationSummary(errors));
+    }
+    Ok(())
+}
+
+fn visit_statement(statement: &Statement) -> Vec<Error> {
+    let mut result = vec![];
+    if statement.identities().is_empty() {
+        result.push(Error::EmptyIdentities);
+    }
+    if statement.operations().is_empty() {
+        result.push(Error::EmptyOperations);
+    }
+    // resources list can be empty only for connect operation.
+    if statement.resources().is_empty() && !is_connect_op(statement) {
+        result.push(Error::EmptyResources);
+    }
+    result
+}
+
+fn visit_identity(value: &str) -> Result<(), Error> {
+    if value.is_empty() {
+        return Err(Error::InvalidIdentity(value.into()));
+    }
+    if let Some(variable) = substituter::extract_variable(value) {
+        if VALID_IDENTITY_VARIABLES.get(variable).is_none() {
+            return Err(Error::InvalidIdentityVariable(variable.into()));
+        }
+    }
+    Ok(())
+}
+
+fn visit_operation(value: &str) -> Result<(), Error> {
+    match value {
+        "mqtt:publish" | "mqtt:subscribe" | "mqtt:connect" => Ok(()),
+        _ => Err(Error::InvalidOperation(value.into())),
+    }
+}
+
+fn visit_resource(value: &str) -> Result<(), Error> {
+    if value.is_empty() {
+        return Err(Error::InvalidResource(value.into()));
+    }
+    if let Some(variable) = substituter::extract_variable(value) {
+        if VALID_RESOURCE_VARIABLES.get(variable).is_none() {
+            return Err(Error::InvalidResourceVariable(variable.into()));
+        }
+    }
+    if TopicFilter::from_str(value).is_err() {
+        return Err(Error::InvalidResource(value.into()));
+    }
+    Ok(())
+}
+
+fn is_connect_op(statement: &Statement) -> bool {
+    statement.operations().len() == 1 && statement.operations()[0] == "mqtt:connect"
+}
+
+lazy_static! {
+    static ref VALID_IDENTITY_VARIABLES: HashSet<String> = HashSet::from_iter(vec![
```

Comment:
I wonder if using small array [&str] will be more performant than HashSet. because of compiler optimizations and data fetching